### PR TITLE
Implement opcode POP

### PR DESCRIPTION
**Pull request by scroll-dev**
rebase to the latest main

### CODE REVIEW

**Review by miha-stopar — CHANGES_REQUESTED** [review]
Looks good to me! I have only two comments: there is some formatting to be fixed as reported by clippy; and there are two TODOs - for some additional constraints and for some failure tests. I think there are no additional constraints for POP and perhaps there is no need for failure tests here as no words are assigned.

File: zkevm-circuits/src/evm_circuit/op_execution/pop.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error> {
+        match execution_step.case {
+            Case::Success => self.assign_success(
+                region,
+                offset,
+                op_execution_state,
+                execution_step,
+            ),
+            Case::StackOverflow => {
```
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/op_execution/pop.rs:168]: Typo

```suggestion
            Case::StackUnderflow => {
```

File: zkevm-circuits/src/evm_circuit/op_execution.rs [hunk 2]
```diff
@@ -17,10 +17,11 @@ use std::{collections::HashMap, ops::Range};
 mod arithmetic;
 mod comparator;
 mod push;
-
+mod pop;
```
Comment (gaswhat) [inline zkevm-circuits/src/evm_circuit/op_execution.rs:20]: add an extra line after this.

File: zkevm-circuits/src/evm_circuit/op_execution.rs [hunk 3]
```diff
@@ -541,7 +545,13 @@ impl<F: FieldExt> OpExecutionGadget<F> {
                     core_state,
                     execution_step,
                 )?,
-                OpcodeId::LT | OpcodeId::GT => self.lt_gadget.assign(
+                  OpcodeId::LT | OpcodeId::GT => self.lt_gadget.assign(
```
Comment (gaswhat) [inline zkevm-circuits/src/evm_circuit/op_execution.rs:548]: indent is wrong here.

File: zkevm-circuits/src/evm_circuit/op_execution/pop.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
+//use crate::evm_circuit::{FixedLookup, Lookup};
```
Comment (gaswhat) [inline zkevm-circuits/src/evm_circuit/op_execution/pop.rs:1]: remove comment

File: zkevm-circuits/src/evm_circuit/op_execution/pop.rs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+impl<F: FieldExt> OpGadget<F> for PopGadget<F> {
+    const RESPONSIBLE_OPCODES: &'static [OpcodeId] = &[
+        OpcodeId::POP // 0x50 of op id
+    ];
+
+
+    const CASE_CONFIGS: &'static [CaseConfig] = &[
+        CaseConfig {
+            case: Case::Success,
+            num_word: 0,// no operand required for pop
```
Comment (gaswhat) [inline zkevm-circuits/src/evm_circuit/op_execution/pop.rs:36]:
```suggestion
            num_word: 0, // no operand required for pop 
```

File: zkevm-circuits/src/evm_circuit/op_execution/pop.rs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+};
+use super::{CaseAllocation, CaseConfig, OpExecutionState, OpGadget};
+use bus_mapping::evm::{GasCost, OpcodeId};
+use halo2::plonk::Error;
+use halo2::{arithmetic::FieldExt, circuit::Region};
+use std::convert::TryInto;
+use crate::util::Expr;
+
+#[derive(Clone, Debug)]
+struct PopSuccessAllocation<F> {
```
Comment (gaswhat) [inline zkevm-circuits/src/evm_circuit/op_execution/pop.rs:13]: If there's only one cell in `PopSuccessAllocation`, we can remove this struct def

File: zkevm-circuits/src/evm_circuit/op_execution/pop.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/op_execution/pop.rs]: fixed

File: zkevm-circuits/src/evm_circuit/op_execution.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/op_execution.rs:20]: added

File: zkevm-circuits/src/evm_circuit/op_execution.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/op_execution.rs]: align it now

File: zkevm-circuits/src/evm_circuit/op_execution/pop.rs (hunk 6 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/op_execution/pop.rs]: removeed

**Comment by DreamWuGit** [comment]
> Looks good to me! I have only two comments: there is some formatting to be fixed as reported by clippy; and there are two TODOs - for some additional constraints and for some failure tests. I think there are no additional constraints for POP and perhaps there is no need for failure tests here as no words are assigned.

I think so and remove them , and clippy fixed

**Review by miha-stopar — APPROVED** [review]
Looks cool to me! Thanks for the PR!

**Review by ChihChengLiang — APPROVED** [review]
Looks good to me. The implementation reflects the [spec](https://github.com/appliedzkp/zkevm-specs/blob/master/specs/stack_ops.md). Can merge once the CI pass

**Comment by ChihChengLiang** [comment]
Thanks you the PR @DreamWuGit!

**Pull request merged**